effectful/handlers/pyro.py

```diff
@@ -252,190 +264,280 @@ def __repr__(self):
         return f"Naming({self.name_to_dim})"
 
 
-class PositionalDistribution(pyro.distributions.torch_distribution.TorchDistribution):
-    """A distribution wrapper that lazily converts indexed dimensions to
-    positional.
+@defop
+def named_distribution(
+    d: Annotated[TorchDistribution, Scoped[A | B]],
+    *names: Annotated[Operation[[], int], Scoped[B]],
+) -> Annotated[TorchDistribution, Scoped[A | B]]:
+    match defterm(d):
+        case Term(op=_call, args=(dist_constr, *args)) if _call is call:
+            if dist_constr is dist.Independent:
+                base_dist, reinterpreted_batch_ndims = args
+                return dist.Independent(
+                    named_distribution(
+                        typing.cast(TorchDistribution, base_dist), *names
+                    ),
+                    reinterpreted_batch_ndims,
+                )
+            else:
+                named_args = []
+                for a in args:
+                    assert isinstance(a, torch.Tensor)
+                    named_args.append(
+                        Indexable(typing.cast(torch.Tensor, a))[
+                            tuple(n() for n in names)
+                        ]
+                    )
+                assert callable(dist_constr)
+                return dist_constr(*named_args)
+        case _:
+            raise NotImplementedError
 
-    """
 
-    indices: Mapping[Operation[[], int], int]
+@defop
+def positional_distribution(
+    d: Annotated[TorchDistribution, Scoped[A]],
+) -> Tuple[TorchDistribution, Naming]:
+    match defterm(d):
+        case Term(op=_call, args=(dist_constr, *args)) if _call is call:
+            if dist_constr is dist.Independent:
+                base_dist, reinterpreted_batch_ndims = args
+                pos_base_dist, naming = positional_distribution(
+                    typing.cast(TorchDistribution, base_dist)
+                )
+                return (
+                    dist.Independent(pos_base_dist, reinterpreted_batch_ndims),
+                    naming,
+                )
+            else:
+                assert callable(dist_constr)
+                base_dist = dist_constr(*args)
+                indices = sizesof(d).keys()
+                n_base = len(base_dist.batch_shape) + len(base_dist.event_shape)
+                naming = Naming.from_shape(indices, n_base)
+                pos_args = [to_tensor(a, indices) for a in args]
+                pos_dist = dist_constr(*pos_args)
+                return pos_dist, naming
+        case _:
+            raise NotImplementedError
+
 
-    def __init__(
-        self, base_dist: pyro.distributions.torch_distribution.TorchDistribution
-    ):
-        self.base_dist = base_dist
-        self.indices = sizesof(base_dist)
+class _DistributionTerm(Term[TorchDistribution], TorchDistribution):
```

Comment:
This should be registered with and accessed through `defdata`, but we can also address that in a followup PR.